### PR TITLE
dune3d: Fix darwin build

### PR DESCRIPTION
**Pull request by afh**
## Description of changes

* Fix darwin build
* Refactor to use `finalAttrs` instead of `rec`
* Refactor to avoid use of `meta = with lib;`

ℹ️ Pinning opencascade-occt to 7.6 helped with building on darwin immensly, are folks okay to downgrade to this version for all platforms or should this be darwin specific?

Edit: Using the latest opencascade-occt failed to build on darwin due to libraries not being found, e.g. `-lTKSTEP`

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by dwt** [comment]
Preliminary testing suggests it works on macOS Sonoma 14.5 (23F79)

File: pkgs/by-name/du/dune3d/package.nix [hunk 1]
```diff
@@ -10,33 +11,48 @@
...
+
   nativeBuildInputs = [
     gobject-introspection
     meson
     ninja
     pkg-config
     wrapGAppsHook3
+    libxml2 # for xmllint
+  ] ++ lib.optionals stdenv.isDarwin [
+    clang_17 # required by use of std::format
```
Comment (jue89) [inline pkgs/by-name/du/dune3d/package.nix:53]: Change to whole stdenv to clang if the package is built on Darwin. (cf. the project's flake.nix)

File: pkgs/by-name/du/dune3d/package.nix [hunk 2]
```diff
@@ -10,33 +11,48 @@
...
+    rev = "v${finalAttrs.version}";
     hash = "sha256-Z/kdOc/MbnnEyRsel3aZGndTAy1eCdAK0Wdta0HxaE4=";
   };
 
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    # uuid.h is bundled with system frameworks on darwin
+    substituteInPlace meson.build \
+      --replace-fail "uuid = dependency('uuid')" "" \
+      --replace-fail "build_dependencies += uuid" ""
+  '';
```
Comment (jue89) [inline pkgs/by-name/du/dune3d/package.nix:43]: Replace the build input `linuuid` with `libossp_uuid` for Darwin.

**Review by jue89 — CHANGES_REQUESTED** [review]
Thanks for your effort!

I don't have a MacBook with up-to-date MacOS at hand - so I can't test your changes.

Please have a look into the project's [flake.nix](https://github.com/dune3d/dune3d/blob/48b24ec90732929148589ab11b354bb2c0497891/flake.nix). It already has support for Darwin and might give you some inspiration.

See my comments down below:

**Comment by afh** [comment]
Thanks for the review and helpful feedback, @jue89, very much appreciated. I went ahead and made the respective changes to this PR.

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 319014` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dune3d</li>
  </ul>
</details>

**Review by jue89 — APPROVED** [review]
> Result of `nixpkgs-review pr 319014` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 
> 1 package built:

Same for x86_64-linux.

LGTM.

**Review by 4evy — APPROVED** [review]

**Comment by afh** [comment]
Thanks for merging this, @siraben, much appreciated 🙂

**Pull request merged**